art/estimators/sequence/tensorflow.py

```diff
@@ -0,0 +1,453 @@
+                  could have different lengths. A possible example of `x` could be:
+                  `x = np.ndarray([[0.1, 0.2, 0.1, 0.4], [0.3, 0.1]])`.
+        :param y: Target values of shape (nb_samples). Each sample in `y` is a string and it may possess different
+                  lengths. A possible example of `y` could be: `y = np.array(['SIXTY ONE', 'HELLO'])`.
+        :param batch_mode: If `True` calculate gradient per batch or otherwise per sequence.
+        :return: Loss gradients of the same shape as `x`.
+        """
+        if batch_mode:
+            gradients = self._loss_gradient_per_batch(x, y)
+        else:
+            gradients = self._loss_gradient_per_sequence(x, y)
+        return gradients
+
+    def _loss_gradient_per_batch(self, x: np.ndarray, y: np.ndarray) -> np.ndarray:
+        """
+        Compute the gradient of the loss function w.r.t. `x` per batch.
+        """
+        import tensorflow.compat.v1 as tf1
+        from lingvo.core.py_utils import NestedMap
+
+        assert x.shape[0] == y.shape[0], "Number of samples in x and y differ."
+
+        # pad input
+        x_padded, mask, mask_frequency = self._pad_audio_input(x)
+
+        # get loss gradients
+        feed_dict = {
+            self._x_padded: x_padded,
+            self._y_target: y,
+            self._mask_frequency: mask_frequency,
+        }
+        gradients_padded = self._sess.run(
+            self._loss_gradient(self._x_padded, self._y_target, self._mask_frequency), feed_dict
+        )
+
+        # undo padding, i.e. change gradients shape from (nb_samples, max_length) to (nb_samples)
+        lengths = mask.sum(axis=1)
+        gradients = list()
+        for gradient_padded, length in zip(gradients_padded, lengths):
+            gradient = gradient_padded[:length]
+            gradients.append(gradient)
+
+        return np.array(gradients)
+
+    def _loss_gradient_per_sequence(self, x: np.ndarray, y: np.ndarray) -> np.ndarray:
+        """
+        Compute the gradient of the loss function w.r.t. `x` per sequence.
+        """
+        import tensorflow.compat.v1 as tf1
+        from lingvo.core.py_utils import NestedMap
+
+        assert x.shape[0] == y.shape[0], "Number of samples in x and y differ."
+
+        gradients = list()
+        for x_i, y_i in zip(x, y):
+            _, _, mask_frequency = self._pad_audio_input(np.array([x_i]))
+            feed_dict = {
+                self._x_padded: np.expand_dims(x_i, 0),
+                self._y_target: np.array([y_i]),
+                self._mask_frequency: mask_frequency,
+            }
+            # get loss gradient
+            gradient = self._sess.run(
+                self._loss_gradient(self._x_padded, self._y_target, self._mask_frequency), feed_dict
+            )
+            gradients.append(np.squeeze(gradient))
+
+        return np.array(gradients)
+
+    def set_learning_phase(self) -> NoReturn:
```

Comment:
The `NoReturn` leads to a `Signature of incompatible with supertype "NeuralNetworkMixin"` error in mypy. Is there a way to keep `NoReturn` for `NotImplementedError` cases but avoid the mpypy error?

Reply:
Thanks for highlighting the `mypy` error. I overlooked that initially.

I think we should use a `Union` in the return type of the `NeuralNetworkMixin`. I have added a note to pick this up myself later. Reverted the type now back to the return type of the abstract method definition (which are technically the wrong types if we raise the `NotImplementedError`...).